packages/insured-bridge-relayer/src/ProfitabilityCalculator.ts

```diff
@@ -0,0 +1,227 @@
+      cumulativeGasPrice: cumulativeGasPrice.toString(),
+      relayerDiscount: fromWei(this.relayerDiscount),
+      slowRevenue: slowRevenue.toString(),
+      speedUpRevenue: speedUpRevenue.toString(),
+      instantRevenue: instantRevenue.toString(),
+    });
+
+    // Calculate the transaction submission costs depending on the gas price and the token type being relayed. This
+    // considers the relayer discount, applied to all relay actions equally.
+    const ethSubmissionCost = this.getRelayEthSubmissionCost(cumulativeGasPrice, tokenType);
+
+    // Calculate the expected revenue, in ETH, based on the amount of tokens being offered in fees.
+    const ethRevenue = this.getEthRevenue(tokenEthPrice, slowRevenue, speedUpRevenue, instantRevenue);
+
+    // Calculate the relay profitability as the difference between revenue and cost.
+    const ethProfitability = this.getProfit(ethSubmissionCost, ethRevenue);
+
+    // Finally, determine the relay submit type based on the profitability.
+    let relaySubmitType: RelaySubmitType;
+    if (
+      ethProfitability.instantEthProfit.gt(toBN(0)) &&
+      ethProfitability.instantEthProfit.gte(ethProfitability.speedUpEthProfit) &&
+      ethProfitability.instantEthProfit.gte(ethProfitability.slowEthProfit)
+    )
+      relaySubmitType = RelaySubmitType.Instant;
+    else if (
+      ethProfitability.speedUpEthProfit.gt(toBN(0)) &&
+      ethProfitability.speedUpEthProfit.gt(ethProfitability.slowEthProfit)
+    )
+      relaySubmitType = RelaySubmitType.SpeedUp;
+    else if (ethProfitability.slowEthProfit.gt(toBN(0))) relaySubmitType = RelaySubmitType.Slow;
+    else relaySubmitType = RelaySubmitType.Ignore;
+
+    this.logger.debug({
+      at: "ProfitabilityCalculator",
+      message: "Considered relay profitability",
+      l1Token,
+      tokenType: TokenType[tokenType],
+      tokenEthPrice: fromWei(tokenEthPrice),
+      cumulativeGasPrice: cumulativeGasPrice.toString(),
+      relayerDiscount: fromWei(this.relayerDiscount),
+      ethSubmissionCost: objectMap(ethSubmissionCost, (value: BN) => fromWei(value)),
+      ethRevenue: objectMap(ethRevenue, (value: BN) => fromWei(value)),
+      ethProfitability: objectMap(ethProfitability, (value: BN) => fromWei(value)),
+      relaySubmitType: RelaySubmitType[relaySubmitType],
+    });
+    return relaySubmitType;
+  }
+
+  getRelayEthSubmissionCost(
+    gasPrice: BN,
+    tokenType: TokenType
+  ): {
+    slowEThCost: BN;
+    speedUpEthCost: BN;
+    instantEthCost: BN;
+  } {
+    const discount = fixedPoint.sub(this.relayerDiscount);
+    switch (tokenType) {
+      case TokenType.WETH:
+        return {
+          slowEThCost: gasPrice.mul(toBN(across.constants.SLOW_ETH_GAS)).mul(discount).div(fixedPoint),
+          speedUpEthCost: gasPrice.mul(toBN(across.constants.SPEED_UP_ETH_GAS)).mul(discount).div(fixedPoint),
+          instantEthCost: gasPrice.mul(toBN(across.constants.FAST_ETH_GAS)).mul(discount).div(fixedPoint),
```

Comment:
just FYI these constants might not be correct, francesco found out that the constants need to be subtracted from each other to get the correct cost, which i didnt realize when i set these constants. See here, hes doing math on the constants: https://github.com/across-protocol/across-frontend/blob/master/src/utils/bridge.ts#L70

Reply:
ye so the subtraction thing that Francesco did we did together IRL. that was to find the cost of the speedup, over and above the slow relay cost. he then later adds these together to get the overall "fast" cost here: https://github.com/across-protocol/across-frontend/blob/fbf8ea1cc21a6c93b8bb9305e760560c371d8810/src/utils/bridge.ts#L101

I dont think this is _strictly_ correct. Ideally, we do no addition in the front end and quote the user (and compute the slow/instant rewards) using the full amount on `FAST_x_GAS` and do no addition. @mrice32 computed these constants by looking at the _cumlative_ gas used over a full relay, including the `relayAndSpeedUp` call and finally the `settleRelay` call later on.

Reply:
This does bring up another point that @mrice32 and myself were discussing earlier on this week. right now, the front end considers the gas price using a diffrent priority fee to what the bots will use. this means that in general the front end will be _under_ quoting the cost of relays, given the bots always use a fast priority fee. we should consider changing this in the front end to more accurately represent what the bots will be paying.

Reply:
got it good to know